File: python/hashing/nameToPhoneNumber.py
from random import *

class NameToPhoneNumber:
    class Pair:
        def __init__(self, key, value):
            self.key = key
            self.value = value

    def __init__(self):
        self.L = 15
        self.p = 10 ** 15 + 100011
        self.m = 1000
        self._setRandomKeys()
        self.chains = [None] * self.m
    
    def _setRandomKeys(self):
        self.x = randint(1, self.p - 1)
        self.a = randint(1, self.p - 1)
        self.b = randint(0, self.p - 1)

    def _polyHash(self, S):
        h = 0
        for i in range(len(S) - 1, -1, -1):
            h = (h * self.x + ord(S[i])) % self.p

        return h

    def _universalFamilyHash(self, x):
        return ((self.a * x + self.b) % self.p) % self.m

    def _hash(self, S):
        return self._universalFamilyHash(self._polyHash(S))

    def add(self, key, value):
        chain = self.chains[self._hash(key)]
        
        if chain is None:
            self.chains[self._hash(key)] = chain = []

        for pair in chain:
            if pair.key == key:
                pair.value = value
        chain.append(self.Pair(key, value))

    def get(self, key):
        chain = self.chains[self._hash(key)]

        if chain is not None:
            for pair in chain:
                if pair.key == key:
                    return pair.value
        return None

File: java/src/hashing/NameToPhoneNumber.java
package hashing;

import java.math.BigInteger;
import java.util.LinkedList;
import java.util.concurrent.ThreadLocalRandom;


public class NameToPhoneNumber {
    private class Pair {
        public String key;

        public String value;

        public Pair(String key, String value) {
            this.key = key;
            this.value = value;
        }
    }

    private int L = 15;

    private long p = (long) Math.pow(10, 15) + 100011;

    private int m  = 1000;

    private BigInteger a;

    private BigInteger b;

    private BigInteger x;

    private LinkedList<Pair>[] chains = new LinkedList[this.m];

    public NameToPhoneNumber() {
        this.randomKeys();
    }

    private void randomKeys() {
        this.x = BigInteger.valueOf(ThreadLocalRandom.current().nextLong(1, p - 1));
        this.a = BigInteger.valueOf(ThreadLocalRandom.current().nextLong(1, p - 1));
        this.b = BigInteger.valueOf(ThreadLocalRandom.current().nextLong(0, p - 1));
    }

    private BigInteger polyHash(String S) {
        BigInteger hash = BigInteger.ZERO;

        for(int i = S.length() - 1; i >= 0 ; i--)
            hash = hash.multiply(this.x).add(BigInteger.valueOf((int)S.charAt(i))).mod(BigInteger.valueOf(this.p));

        return hash;
    }

    private int universalFamilyHash(BigInteger x) {
        return this.a.multiply(this.x).add(this.b).mod(BigInteger.valueOf(this.p)).mod(BigInteger.valueOf(this.m)).intValue();
    }

    private int hash(String S) {
        return this.universalFamilyHash(this.polyHash(S));
    }

    public void add(String key, String value) {
        LinkedList<Pair> chain = this.chains[this.hash(key)];

        if (chain == null)
            chain = new LinkedList<Pair>();

        Pair pair = chain.stream().filter(x -> key.equals(x.key)).findAny().orElse(null);
        if (pair != null) {
            pair.value = value;
        } else {
            chain.add(new Pair(key, value));
            this.chains[this.hash(key)] = chain;
        }
    }

    public String get(String key) {
        LinkedList<Pair> chain = this.chains[this.hash(key)];

        if (chain == null)
            return null;

        return chain.stream().filter(x -> key.equals(x.key)).map(x -> x.value).findAny().orElse(null);
    }
}
